### ISSUE TITLE
[launch] deal with very long cloud-init times

### ISSUE THREAD

**Issue by Saviq** [canonical/multipass#1039]
With a custom `multipass launch --cloud-init` that runs over the allocated 5 minutes, we spit out a bad error:

```
launch failed: The following errors occurred:
timed out waiting for initialization to complete
```

The launch didn't fail, it's just that we got bored waiting for cloud-init to complete.

We need to deal with that better - maybe `--cloud-init-timeout`, or maybe we just need to give up a better message, suggesting the use of `multipass exec <instance> -- cloud-init status --wait`.

**Comment by townsend2010** [comment]
After a conversation with @matthewpaulthomas, we discussed if `launch` is running in a non-interactive session, we just wait until `cloud-init` finishes, no matter how long it takes.

For an interactive session, we wait for some arbitrary amount time (like now), but instead of failing the launch, we ask if the user if they want to continue waiting for "initialization".  If they don't want to wait, then `launch` will exit successfully.  Of course, if the user wants to continue waiting, then it will keep running the spinner until it finishes.  

And if `cloud-init` finishes while the question is being posed and the user has not answered yet, then `launch` will just exit with success.

**Comment by matthewpaulthomas** [comment]
In our meeting yesterday I suggested:
- If Multipass is running non-interactively (in a script), don’t have a timeout.
- If Multipass is running interactively, print and update something like:
  ```(cloud-init has taken 1:00 so far, press Esc to connect without waiting)```

**Comment by ricab** [comment]
Hi @matthewpaulthomas. I like the approach you propose. I just discussed it a bit further with @townsend2010 and wanted to address some details on your second point:
- if we allow the user not to wait for cloud-init, I don't think we should make them wait to get that option;
- `connect` is not really what happens when launch finishes (I think you had `shell` in your mind);
- to be fair to cloud-init, we should probably not mention it, as it is merely performing requested initializations (they requested that in an earlier issue).

So, adapting it a bit, perhaps the following spinner message could shown during the cloud-init step: `finishing initialization (press Esc to continue) [spinner]`.

**Comment by Saviq** [comment]
I wonder if we could also allow another key to increase verbosity, in which case we'd forward `cloud-init.log`'s output to the user.

**Comment by matthewpaulthomas** [comment]
I think “`press Esc to continue`” would be misleading, seeming as if it is paused waiting for a response.

> I wonder if we could also allow another key to increase verbosity, in which case we'd forward `cloud-init.log`'s output to the user.

So how about:
```(Still going… Type V for details, Ctrl C to cancel, or Esc to start shell without waiting)```

**Comment by ricab** [comment]
> `press Esc to connect without waiting`
> `Esc to start shell without waiting`

`multipass launch --cloud-init` does not start a shell. Am I missing something?

**Comment by matthewpaulthomas** [comment]
> `press Esc to connect without waiting`

I suggested this because “connect” was the word I thought @Saviq used in the meeting.

> `Esc to start shell without waiting`

I suggested this because you wrote “I think you had `shell` in your mind”.

If neither of those are correct, what _would_ happen if you pressed Esc?

**Comment by ricab** [comment]
> I suggested this because you wrote “I think you had `shell` in your mind”.

Yes, I was referring to the command `shell`. Since you said "`connect`", I thought you might be confusing the behavior of command `launch` with that of command `shell`. If this issue was about `shell`, it would indeed be totally appropriate to say either "connect" or "shell" in the message.

> what would happen if you pressed Esc?

I suppose the same that would happen when the wait finished: return to the calling shell.

**Comment by matthewpaulthomas** [comment]
So it would be the equivalent of (in bash) Ctrl Z then `bg`? If so:

`(Still going… V for details, Ctrl C to cancel, or Esc to continue in background)`

**Comment by ricab** [comment]
Similar: no real job left running in the calling shell (things are happening inside the instance), but yeah, same idea.

**Comment by ricab** [comment]
@matthewpaulthomas I like the idea of providing an informative view of what is happening followed by interactive options. I have some questions on that particular msg and whether it could be improved further.

> `Still going`

We can't use "cloud-init" but maybe we could say something like `Customizing instance...`? 

We already update the msg with multiple steps (e.g. "downloading image", "extracting image", etc.). Describing what we're doing would be coherent. It would also justify the new ability to cancel. Otherwise, the user may not understand why we make them wait before we allow them to stop waiting.

> `V for details`

Nice, sounds good to me.

> `Ctrl C to cancel`

Would we have to add this to all previous messages? What aboud other commands? `Ctrl C` is not particular to multipass, so should we instruct the user on how to use the shell?

Incidently, cancelling launches has not always been the best experience, although that is more of a dev matter :slightly_smiling_face: 

> B to continue in background

I personally preferred your earlier suggestion with `Esc`. I find it more intuitive, and familiar to muscle memory.

**Comment by matthewpaulthomas** [comment]
> We can't use "cloud-init" but maybe we could say something like `Customizing instance...`?

[The CLI spec](https://docs.google.com/document/d/1NW53qrqJDCfjOgpcwoptsO9-L1sO-IInD-dIKEkpGJc/edit#heading=h.nezo34efmyc4) says “`Waiting for cloud-init to complete`”. Is it out of date?

I included “Still” under the impression that this line would be below that one — and that even if common, it would not be guaranteed feedback, since it wouldn’t appear if the next launch is quick or has no cloud-init at all. (Compare “Still loading…” when a Google Sheets spreadsheet takes a long time to open.) But if it’s part of a single cloud-init line, so it _will_ always appear, “Still” isn’t necessary.

After that I was trying to keep it to 80 characters so that it would fit on one line.

> We already update the msg with multiple steps (e.g. "downloading image", "extracting image", etc.). Describing what we're doing would be coherent. It would also justify the new ability to cancel.

Sure, downloading, extracting etc could take minutes too, so both cancelling and backgrounding would be useful from the beginning. On the other hand, advertising them at the beginning of every launch would be noisy and pessimistic. Maybe these tips should appear during whichever step — if any — is the first to take more than _n_ seconds?

>> `Ctrl C to cancel`
>
> Would we have to add this to all previous messages? What aboud other commands? Ctrl C is not particular to multipass, so should we instruct the user on how to use the shell?

I understand this is weird. With my limited command-line experience, it seemed less bad than the alternatives:
- We don’t give any hint of how to cancel. Some small fraction won’t know how. Another fraction will know, but will worry that it will leave the instance in a broken state.
- We add our own cancel command. Extra work to remember, and some people will wonder what the difference is between that and Ctrl C.

What other factors are there?

> > `B to continue in background`
>
> I personally preferred your earlier suggestion with Esc. I find it more intuitive, and familiar to muscle memory.

What other commands use Esc to _background_ something, as opposed to cancelling it or skipping a step?

**Comment by ricab** [comment]
> The CLI spec says “Waiting for cloud-init to complete”. Is it out of date?

My understanding is that we were asked not to mention cloud-init, to avoid giving the impression it was slow when it takes as long as the operations requested by the user. Perhaps others can clarify?

> [...] under the impression that this line would be below that one [...]

We use only one line, whose text is replaced as things progress. Here is an example of [what we get out of `multipass launch`](https://asciinema.org/a/BpJEO58IfbgGcofP3Anlm9yt3) today.

> [...] both cancelling and backgrounding would be useful from the beginning [...]

Cancelling I believe we support already (in the sense that, if we don't behave well on ctrl+c, I think that classifies as a bug). "Backgrounding" the rest of the launch procedure would require more effort than might be apparent at first. More so if that is extended to other commands. OTH some coinciding work may already be planned (decoupling client/daemon for tests).

> [...] noisy and pessimistic [...]

Good point.

> Maybe these tips should appear during whichever step — if any — is the first to take more than n seconds?

Yeah, I like that better. The user has the option to stop waiting all along, it is just that the tips come later, right? They would know about their options from then on. I think that is a nice balance.

>  [...] Ctrl C [...]

I see. OK, I think I understand your reasoning better now. So we'd include Ctrl+C in the hints whenever we provided them, but that wouldn't happen all the time, correct?

> What other commands use Esc to background something[...]?

Something that comes to mind is IDE's building code: when pressing Esc you'd escape the build pane, but not stop the build itself. Or for instance sublime's console: pressing Esc will get you out but things happening there will continue. The console pane itself gets closed, just as the multipass client would. It is not *really* being put into background as would happen with Ctrl+Z.

**Comment by matthewpaulthomas** [comment]
Thanks for the recording! My lack of testing is showing — I’d thought all those messages were on separate lines. That changes my view of how they should be structured: for example, I’d put the spinner at the start of the line so it doesn’t jump around from step to step.

> So we'd include Ctrl+C in the hints whenever we provided them, but that wouldn't happen all the time, correct?

Right. That would introduce a new problem: if the hints appeared at a step earlier than the cloud-init one, then you’d never learn that that particular step had an extra feature (show details).

But if it’s all on one line, we don’t need to worry about which step to introduce the hints at! We can introduce them and persist them, right-aligned and therefore not jumping around, whenever _the whole process_ has taken longer than _n_ seconds. For example (underscores here just to avoid GitHub collapsing the spaces):

`/ Retrieving image: 0%`
…
`— Retrieving image: 63% _____ (Ctrl C to cancel, Esc to continue in background)`
…
`\ Initializing (V for details, Ctrl C to cancel, Esc to continue in background)`

**Comment by Saviq** [comment]
One thing came to mind: when we introduce such a behaviour, we'll need to also introduce a way for the user to introspect and interrupt the operations, as a "runaway" process may run indefinitely.

**Comment by dominikborkowski** [comment]
Just adding to the discussion: having an option to change the timeout would be great. As it stands prolonged package installs with cloud-init result in the timeout error, which may or may not be indicative of an actual problem. 

For people who are using multipass to debug & polish their cloud-init configs having such option would eliminate needless debugging.

Thanks!

**Comment by indrayam** [comment]
I could not agree more with @dominikborkowski. I was excited about getting multipass to be my tool of choice to spin a quick and custom Bionic-based VM. Unfortunately, my startup keeps timing out after working just fine earlier today.

Now, instead of getting a good error message, I am trying to figure out what do I need to change (or take out) to make things work.

👎

**Comment by townsend2010** [comment]
Hi @indrayam,

Is it timing out at the `Waiting for initialization` stage of `launch`?  If so, then that only indicates that `cloud-init` hasn't finished, but the instance is up and running.  Depending on the order of `cloud-init` and if the user and ssh keys are set up, you could even `multipass shell` using a different terminal.

We will work on a solution for this, it's just not the highest priority since the instance is regarded as functional at this point.

**Comment by frederikbosch** [comment]
@townsend2010 Yes it is. Maybe a simple environment variable to do the trick? `CLOUD_INIT_TIMEOUT=10000 multipass launch`?

Sidenote: nice project, like it so far!

**Comment by missmah** [comment]
A year later and no activity here? This is frustratingly killing my `cloud-init`, and making `multipass` useless to me (I thought it would be very useful for testing `cloud-init` `yaml` before deploying to services like `Azure`).

**Comment by Saviq** [comment]
Hi @missmah, we have it on our immediate roadmap, milestoned ~end of April.

**Comment by imakan** [comment]
anyone can tell me how to resolve? 
`timeout initialization to complete`

**Comment by Saviq** [comment]
@imakan you should be able to `multipass shell` inside the instance to see in `/var/log/cloud*` for what's going on. If it just took too long, that's exactly what this issue is about.

**Comment by imakan** [comment]
thank you，As you said。the logs tell me no mistakes。It's a long wait

**Comment by michaelschem** [comment]
Is there a solution to this yet?  Very frustrating to not have the ability to configure the timeout.

**Comment by dmuth** [comment]
Hi,

I'm about 10 minutes into my first ever foray with multipass and found this issue via Google.

I'm working through [the tutorial](https://multipass.run/) and tried launching an instance with `multipass launch -n bar --cloud-init cloud-config.yaml` and this relatively sample `cloud-init.yml`:

```
groups:
  - ubuntu: [root,sys]
  - cloud-users
```

Seems straightforward--maybe I have a typo in that maybe not?  Problem is, I don't really know because there's no useful output and some Google searches for things like `multipass view console` weren't much help to me--I've been watching my instance stuck on `Waiting for initialization to complete` for the last 5 minutes and it _just_ timed out as I was typing this.

There's no way that it took 5 minutes to create two groups.

Further up in this thread, I see some really good suggestions from others, and I would urge the developers to consider implementing them, because right now it's very difficult to try and troubleshoot when things go wrong in multipass, and I'm sour I'm not the only user who has had my perception of the product soured.

-- Doug

**Comment by Saviq** [comment]
@dmuth when an instance is "Waiting for initialization...", you can already access it via `multipass shell ...` and debug there. Look into `/var/log/cloud-init*log` for problems.

**Comment by dmuth** [comment]
Hey @Saviq, that was super helpful, thanks!

I'd like to ask if it's possible to add something like that to the startup process to help other people out who are new to Multipass.  I envision something like:

```
[multipass] $ multipass launch -n bar --cpus 2 --mem 2147483648 --cloud-init ./cloud-config.yml
Waiting for cloud-init to complete (Run multipass shell and check /var/log/cloud-init* for status)
```

Are y'all accepting PRs?  I'd be happy to create one if that'll help!

Thanks,

-- Doug

**Comment by cloudymax** [comment]
Can we at least get an event broadcast when the ip address becomes known or something? waiting on the provisioning to timeout negates any time I saved using kvm and make sit impossible to recommend as a production tool

**Comment by Saviq** [comment]
@cloudymax hey, if you want to get into the instance as soon as it's available, you can interrupt the launch when it reaches "Waiting for initialization…". You can also `multipass start/shell/exec` in a separate terminal, those won't wait for cloud-init to complete.

**Comment by cloudymax** [comment]
> 

this assumed a human user and disregards automation use cases

**Comment by cloudymax** [comment]
> @cloudymax hey, if you want to get into the instance as soon as it's available, you can interrupt the launch when it reaches "Waiting for initialization…". You can also `multipass start/shell/exec` in a separate terminal, those won't wait for cloud-init to complete.
> 
> Also, there's a `--timeout` option now that lets you extend the cloud-init timeout if you have a complex init story.

this assumes human users and disregards automation use cases sensitive to return code variations

**Comment by Saviq** [comment]
@cloudymax that surely should not timeout due to complexity…

This bit looks suspicious, though:
```
groups:
  - ubuntu: [root,sys]
```

Is that correct cloud-init? IIUC [`groups` should be a list](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#users-and-groups)?

And yes, we don't have a good story for dealing with these kinds of issues. The error in the cloud-init data means that we never get to access the instance, as our user isn't provisioned.

**Comment by cloudymax** [comment]
> @cloudymax that surely should not timeout due to complexity…
> 
> This bit looks suspicious, though:
> 
> ```yaml
> groups:
>   - ubuntu: [root,sys]
> ```
> 
> Is that correct cloud-init? IIUC [`groups` should be a list](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#users-and-groups)?
> 
> And yes, we don't have a good story for dealing with these kinds of issues. The error in the cloud-init data means that we never get to access the instance, as our user isn't provisioned.
> 
> > https://multipass.run/docs/launch-command is out of date and does not list the --timeout flag
> 
> Thanks for highlighting that, I've fixed that now - will sync as soon as cache on it expires.

thanks, appreciate the help. I'll mess with it and see if I can get a fix.

workaround I'm using for now with a 5 second timeout:

```zsh
set_vm_ip(){
# grab the new VM's IP
  VM_IP=$(multipass list |grep "${VM_NAME}" |awk '{print $3}')

  while [ $VM_IP == "--" ]
  do
    sleep 2
    multipass start $VM_NAME
    VM_IP=$(multipass list |grep "${VM_NAME}" |awk '{print $3}')
  done

  multipass exec -vvvv $VM_NAME -- sudo ufw allow $SSH_PORT/tcp
  multipass exec -vvvv $VM_NAME -- sudo systemctl reload sshd
}
```

**Comment by mturnaviotov** [comment]
i also have a issues in upgrade packages and install kernel subiquty steps due to small apt timeout and retries count - may be we need to add some settings here with apt timeouts?

Last week my user-data was executed without any problem, but now I can't install kernel
and also should switch to install all packages via late-step block, so this means that packages update, upgrade and install block have a no sense.
also - we have a already provided kernels and images on our HTTP or (t)ftp resource, so may be we need to use it, and just updates if something goes wrong with network?
@matthewpaulthomas, @townsend2010 - what are you thing about it

**Comment by tobe2098** [comment]
This could also be related to other async commands like `stop`(#1044) and `suspend`(#1036). If `launch` is made async:
1. Timeouts will not be necessary unless we want to avoid a blocked thread due to something going wrong inside the VM.
2. The IP should be visible doing an async `list`.
